stacks-signer/src/runloop.rs

```diff
@@ -0,0 +1,439 @@
+                }
+                self.state = State::Dkg;
+            }
+            (RunLoopCommand::DkgSign { message }, State::DkgDone)
+            | (RunLoopCommand::Sign { message }, State::Idle) => {
+                info!("Signing message");
+                if let Ok(msg) = self.coordinator.start_signing_message(message) {
+                    let _ = self
+                        .stacks_client
+                        .send_message(self.signing_round.signer.signer_id, msg);
+                }
+                self.state = State::Sign
+            }
+            (RunLoopCommand::DkgSign { message }, State::SignDone)
+            | (RunLoopCommand::Sign { message }, State::SignDone) => {
+                info!("Done Signing Message: {:?}", message);
+                self.state = State::Exit;
+            }
+            _ => {}
+        }
+    }
+
+    /// Process the event as both a signer and a coordinator
+    fn process_event(
+        &mut self,
+        event: &StackerDBChunksEvent,
+    ) -> (Vec<Message>, Vec<OperationResult>) {
+        //info!("Processing event: {:?}", event);
+        // Determine the current coordinator id and public key for verification
+        let (coordinator_id, coordinator_public_key) =
+            calculate_coordinator(&self.signing_round.public_keys);
+        // Filter out invalid messages
+        let inbound_messages: Vec<Message> = event
+            .modified_slots
+            .iter()
+            .filter_map(|chunk| {
+                let message = bincode::deserialize::<Message>(&chunk.data).ok()?;
+                if verify_msg(
+                    &message,
+                    &self.signing_round.public_keys,
+                    coordinator_public_key,
+                ) {
+                    Some(message)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        // First process all messages as a signer
+        let mut outbound_messages =
+            process_inbound_messages(&mut self.signing_round, inbound_messages.clone())
+                .unwrap_or_default();
+        // If the signer is the coordinator, then next process the message as the coordinator
+        let (messages, results) = if self.signing_round.signer.signer_id == coordinator_id {
+            self.coordinator
+                .process_inbound_messages(inbound_messages)
+                .unwrap_or_default()
+        } else {
+            (vec![], vec![])
+        };
+        outbound_messages.extend(messages);
+        (outbound_messages, results)
+    }
+}
+
+impl RunLoop<FrostCoordinator> {
+    /// Creates new runloop from a config and command
+    pub fn new(config: &Config, command: RunLoopCommand) -> Self {
+        // TODO: should this be a config option?
+        // The threshold is 70% of the shares
+        let threshold = (config.signer_ids_public_keys.key_ids.len() / 10 * 7)
```

Comment:
Should this be `(config.signer_ids_public_keys.key_ids.len() * 7) / 10`?

Reply:
I certainly think so.  Hopfully clippy won't hate the extra parens.

Fixed.